url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/terminal/text/ANSITerminal.java

```diff
@@ -0,0 +1,244 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.terminal.text;
+
+import com.googlecode.lanterna.input.CommonProfile;
+import com.googlecode.lanterna.terminal.TerminalSize;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+
+/**
+ * A common ANSI text terminal implementation
+ * @see <a href="http://en.wikipedia.org/wiki/ANSI_escape_code">Wikipedia</a>
+ * @author Martin
+ */
+public abstract class ANSITerminal extends StreamBasedTerminal
+{
+    ANSITerminal(InputStream terminalInput, OutputStream terminalOutput, Charset terminalCharset)
+    {
+        super(terminalInput, terminalOutput, terminalCharset);
+        addInputProfile(new CommonProfile());
+    }
+
+    private void CSI()
+    {
+        writeToTerminal((byte)0x1b, (byte)'[');
+    }
+    
+    @Deprecated
+    @Override
+    public TerminalSize queryTerminalSize()
+    {        
+        synchronized(writerMutex) {
+            saveCursorPosition();
+            moveCursor(5000, 5000);
+            reportPosition();
+            restoreCursorPosition();
+        }
+        return getLastKnownSize();
+    }
+
+    @Override
+    public TerminalSize getTerminalSize() {
+        queryTerminalSize();
+        return waitForTerminalSizeReport(1000); //Wait 1 second for the terminal size report to come, is this reasonable?
+    }
+
+    @Override
+    public void applyBackgroundColor(Color color)
+    {
+        synchronized(writerMutex) {
+            CSI();
+            writeToTerminal((byte)'4', (byte)((color.getIndex() + "").charAt(0)), (byte)'m');
+        }
+    }
+
+    @Override
+    public void applyBackgroundColor(int r, int g, int b) {
+        if(r < 0 || r > 255)
+            throw new IllegalArgumentException("applyForegroundColor: r is outside of valid range (0-255)");
+        if(g < 0 || g > 255)
+            throw new IllegalArgumentException("applyForegroundColor: g is outside of valid range (0-255)");
+        if(b < 0 || b > 255)
+            throw new IllegalArgumentException("applyForegroundColor: b is outside of valid range (0-255)");
+        
+        synchronized(writerMutex) {
+            CSI();
+            String asString = "48;2;" + r + ";" + g + ";" + b + "m";
+            for(int i = 0; i < asString.length(); i++)
+                writeToTerminal((byte)asString.charAt(i));
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 76:88, 102:114.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.